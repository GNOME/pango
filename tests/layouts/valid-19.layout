{
  "context" : {
    "font" : "serif 12",
    "language" : "en-us",
    "base-gravity" : "south",
    "gravity-hint" : "natural",
    "base-dir" : "weak-ltr",
    "round-glyph-positions" : true,
    "transform" : [
      1,
      0,
      0,
      1,
      0,
      0
    ]
  },
  "comment" : "Emoji",
  "text" : "üòäÔ∏é üòäÔ∏è üá©üá™ ‚úä ‚úãüèæ 0 # üè¥Û†ÅµÛ†Å≥Û†Å£Û†Å°Û†Åø ¬©",
  "font" : "Cantarell 11",
  "output" : {
    "is-wrapped" : false,
    "is-ellipsized" : false,
    "unknown-glyphs" : 0,
    "width" : 166912,
    "height" : 19456,
    "log-attrs" : [
      {
        "char-break" : true,
        "cursor-position" : true,
        "sentence-boundary" : true,
        "sentence-start" : true,
        "backspace-deletes-character" : true,
        "word-boundary" : true
      },
      {},
      {
        "char-break" : true,
        "white" : true,
        "cursor-position" : true,
        "expandable-space" : true,
        "word-boundary" : true
      },
      {
        "line-break" : true,
        "char-break" : true,
        "cursor-position" : true,
        "word-boundary" : true
      },
      {},
      {
        "char-break" : true,
        "white" : true,
        "cursor-position" : true,
        "expandable-space" : true,
        "word-boundary" : true
      },
      {
        "line-break" : true,
        "char-break" : true,
        "cursor-position" : true,
        "word-boundary" : true
      },
      {},
      {
        "char-break" : true,
        "white" : true,
        "cursor-position" : true,
        "expandable-space" : true,
        "word-boundary" : true
      },
      {
        "line-break" : true,
        "char-break" : true,
        "cursor-position" : true,
        "word-boundary" : true
      },
      {
        "char-break" : true,
        "white" : true,
        "cursor-position" : true,
        "expandable-space" : true,
        "word-boundary" : true
      },
      {
        "line-break" : true,
        "char-break" : true,
        "cursor-position" : true,
        "word-boundary" : true
      },
      {},
      {
        "char-break" : true,
        "white" : true,
        "cursor-position" : true,
        "expandable-space" : true,
        "word-boundary" : true
      },
      {
        "line-break" : true,
        "char-break" : true,
        "cursor-position" : true,
        "word-start" : true,
        "word-boundary" : true
      },
      {
        "char-break" : true,
        "white" : true,
        "cursor-position" : true,
        "word-end" : true,
        "expandable-space" : true,
        "word-boundary" : true
      },
      {
        "line-break" : true,
        "char-break" : true,
        "cursor-position" : true,
        "word-boundary" : true
      },
      {
        "char-break" : true,
        "white" : true,
        "cursor-position" : true,
        "expandable-space" : true,
        "word-boundary" : true
      },
      {
        "line-break" : true,
        "char-break" : true,
        "cursor-position" : true,
        "word-boundary" : true
      },
      {},
      {},
      {},
      {},
      {},
      {
        "char-break" : true,
        "white" : true,
        "cursor-position" : true,
        "backspace-deletes-character" : true,
        "expandable-space" : true,
        "word-boundary" : true
      },
      {
        "line-break" : true,
        "char-break" : true,
        "cursor-position" : true,
        "word-boundary" : true
      },
      {
        "line-break" : true,
        "mandatory-break" : true,
        "char-break" : true,
        "white" : true,
        "cursor-position" : true,
        "sentence-boundary" : true,
        "sentence-end" : true,
        "word-boundary" : true
      }
    ],
    "lines" : [
      {
        "start-index" : 0,
        "length" : 68,
        "paragraph-start" : true,
        "direction" : "ltr",
        "runs" : [
          {
            "offset" : 0,
            "length" : 8,
            "text" : "üòäÔ∏é ",
            "bidi-level" : 0,
            "gravity" : "south",
            "language" : "en-us",
            "script" : "common",
            "font" : {
              "description" : "DejaVu Sans 11",
              "checksum" : "69ccd07023a72ceb27a5e5c22f728627353b60a198170f5e58dd7014221abf01",
              "matrix" : [
                1,
                -0,
                -0,
                1,
                0,
                0
              ]
            },
            "flags" : 0,
            "y-offset" : 0,
            "start-x-offset" : 0,
            "end-x-offset" : 0,
            "glyphs" : [
              {
                "glyph" : 5867,
                "width" : 15360,
                "is-cluster-start" : true,
                "log-cluster" : 0
              },
              {
                "glyph" : 268435455,
                "width" : 0,
                "is-cluster-start" : true,
                "log-cluster" : 4
              },
              {
                "glyph" : 3,
                "width" : 5120,
                "is-cluster-start" : true,
                "log-cluster" : 7
              }
            ]
          },
          {
            "offset" : 8,
            "length" : 7,
            "text" : "üòäÔ∏è",
            "bidi-level" : 0,
            "gravity" : "south",
            "language" : "en-us",
            "script" : "common",
            "font" : {
              "description" : "Noto Color Emoji 11",
              "checksum" : "054c1c2e2ca956b920262840dcad39dcf27bb88d6f70b892b2b1180de2c2ab47",
              "matrix" : [
                0.13455956135321101,
                -0,
                -0,
                0.13455956135321101,
                0,
                0
              ]
            },
            "flags" : 0,
            "y-offset" : 0,
            "start-x-offset" : 0,
            "end-x-offset" : 0,
            "glyphs" : [
              {
                "glyph" : 13,
                "width" : 18432,
                "is-cluster-start" : true,
                "is-color" : true,
                "log-cluster" : 0
              },
              {
                "glyph" : 268435455,
                "width" : 0,
                "is-cluster-start" : true,
                "log-cluster" : 4
              }
            ]
          },
          {
            "offset" : 15,
            "length" : 1,
            "text" : " ",
            "bidi-level" : 0,
            "gravity" : "south",
            "language" : "en-us",
            "script" : "common",
            "font" : {
              "description" : "Cantarell 11",
              "checksum" : "5bcb6ee14ee9d210b2e91d643de1fe456e9d1aea770983fdb05951545efebbe2",
              "variations" : {
                "wght" : 0
              },
              "matrix" : [
                1,
                -0,
                -0,
                1,
                0,
                0
              ]
            },
            "flags" : 0,
            "y-offset" : 0,
            "start-x-offset" : 0,
            "end-x-offset" : 0,
            "glyphs" : [
              {
                "glyph" : 1109,
                "width" : 3072,
                "is-cluster-start" : true,
                "log-cluster" : 0
              }
            ]
          },
          {
            "offset" : 16,
            "length" : 8,
            "text" : "üá©üá™",
            "bidi-level" : 0,
            "gravity" : "south",
            "language" : "en-us",
            "script" : "common",
            "font" : {
              "description" : "Noto Color Emoji 11",
              "checksum" : "054c1c2e2ca956b920262840dcad39dcf27bb88d6f70b892b2b1180de2c2ab47",
              "matrix" : [
                0.13455956135321101,
                -0,
                -0,
                0.13455956135321101,
                0,
                0
              ]
            },
            "flags" : 0,
            "y-offset" : 0,
            "start-x-offset" : 0,
            "end-x-offset" : 0,
            "glyphs" : [
              {
                "glyph" : 22,
                "width" : 18432,
                "is-cluster-start" : true,
                "is-color" : true,
                "log-cluster" : 0
              }
            ]
          },
          {
            "offset" : 24,
            "length" : 1,
            "text" : " ",
            "bidi-level" : 0,
            "gravity" : "south",
            "language" : "en-us",
            "script" : "common",
            "font" : {
              "description" : "Cantarell 11",
              "checksum" : "5bcb6ee14ee9d210b2e91d643de1fe456e9d1aea770983fdb05951545efebbe2",
              "variations" : {
                "wght" : 0
              },
              "matrix" : [
                1,
                -0,
                -0,
                1,
                0,
                0
              ]
            },
            "flags" : 0,
            "y-offset" : 0,
            "start-x-offset" : 0,
            "end-x-offset" : 0,
            "glyphs" : [
              {
                "glyph" : 1109,
                "width" : 3072,
                "is-cluster-start" : true,
                "log-cluster" : 0
              }
            ]
          },
          {
            "offset" : 25,
            "length" : 3,
            "text" : "‚úä",
            "bidi-level" : 0,
            "gravity" : "south",
            "language" : "en-us",
            "script" : "common",
            "font" : {
              "description" : "Noto Color Emoji 11",
              "checksum" : "054c1c2e2ca956b920262840dcad39dcf27bb88d6f70b892b2b1180de2c2ab47",
              "matrix" : [
                0.13455956135321101,
                -0,
                -0,
                0.13455956135321101,
                0,
                0
              ]
            },
            "flags" : 0,
            "y-offset" : 0,
            "start-x-offset" : 0,
            "end-x-offset" : 0,
            "glyphs" : [
              {
                "glyph" : 6,
                "width" : 18432,
                "is-cluster-start" : true,
                "is-color" : true,
                "log-cluster" : 0
              }
            ]
          },
          {
            "offset" : 28,
            "length" : 1,
            "text" : " ",
            "bidi-level" : 0,
            "gravity" : "south",
            "language" : "en-us",
            "script" : "common",
            "font" : {
              "description" : "Cantarell 11",
              "checksum" : "5bcb6ee14ee9d210b2e91d643de1fe456e9d1aea770983fdb05951545efebbe2",
              "variations" : {
                "wght" : 0
              },
              "matrix" : [
                1,
                -0,
                -0,
                1,
                0,
                0
              ]
            },
            "flags" : 0,
            "y-offset" : 0,
            "start-x-offset" : 0,
            "end-x-offset" : 0,
            "glyphs" : [
              {
                "glyph" : 1109,
                "width" : 3072,
                "is-cluster-start" : true,
                "log-cluster" : 0
              }
            ]
          },
          {
            "offset" : 29,
            "length" : 7,
            "text" : "‚úãüèæ",
            "bidi-level" : 0,
            "gravity" : "south",
            "language" : "en-us",
            "script" : "common",
            "font" : {
              "description" : "Noto Color Emoji 11",
              "checksum" : "054c1c2e2ca956b920262840dcad39dcf27bb88d6f70b892b2b1180de2c2ab47",
              "matrix" : [
                0.13455956135321101,
                -0,
                -0,
                0.13455956135321101,
                0,
                0
              ]
            },
            "flags" : 0,
            "y-offset" : 0,
            "start-x-offset" : 0,
            "end-x-offset" : 0,
            "glyphs" : [
              {
                "glyph" : 21,
                "width" : 18432,
                "is-cluster-start" : true,
                "is-color" : true,
                "log-cluster" : 0
              }
            ]
          },
          {
            "offset" : 36,
            "length" : 5,
            "text" : " 0 # ",
            "bidi-level" : 0,
            "gravity" : "south",
            "language" : "en-us",
            "script" : "common",
            "font" : {
              "description" : "Cantarell 11",
              "checksum" : "5bcb6ee14ee9d210b2e91d643de1fe456e9d1aea770983fdb05951545efebbe2",
              "variations" : {
                "wght" : 0
              },
              "matrix" : [
                1,
                -0,
                -0,
                1,
                0,
                0
              ]
            },
            "flags" : 0,
            "y-offset" : 0,
            "start-x-offset" : 0,
            "end-x-offset" : 0,
            "glyphs" : [
              {
                "glyph" : 1109,
                "width" : 3072,
                "is-cluster-start" : true,
                "log-cluster" : 0
              },
              {
                "glyph" : 964,
                "width" : 9216,
                "is-cluster-start" : true,
                "log-cluster" : 1
              },
              {
                "glyph" : 1109,
                "width" : 3072,
                "is-cluster-start" : true,
                "log-cluster" : 2
              },
              {
                "glyph" : 1070,
                "width" : 10240,
                "is-cluster-start" : true,
                "log-cluster" : 3
              },
              {
                "glyph" : 1109,
                "width" : 3072,
                "is-cluster-start" : true,
                "log-cluster" : 4
              }
            ]
          },
          {
            "offset" : 41,
            "length" : 24,
            "text" : "üè¥Û†ÅµÛ†Å≥Û†Å£Û†Å°Û†Åø",
            "bidi-level" : 0,
            "gravity" : "south",
            "language" : "en-us",
            "script" : "common",
            "font" : {
              "description" : "Noto Color Emoji 11",
              "checksum" : "054c1c2e2ca956b920262840dcad39dcf27bb88d6f70b892b2b1180de2c2ab47",
              "matrix" : [
                0.13455956135321101,
                -0,
                -0,
                0.13455956135321101,
                0,
                0
              ]
            },
            "flags" : 0,
            "y-offset" : 0,
            "start-x-offset" : 0,
            "end-x-offset" : 0,
            "glyphs" : [
              {
                "glyph" : 10,
                "width" : 18432,
                "is-cluster-start" : true,
                "is-color" : true,
                "log-cluster" : 0
              },
              {
                "glyph" : 268435455,
                "width" : 0,
                "is-cluster-start" : true,
                "log-cluster" : 16
              },
              {
                "glyph" : 268435455,
                "width" : 0,
                "is-cluster-start" : true,
                "log-cluster" : 20
              }
            ]
          },
          {
            "offset" : 65,
            "length" : 3,
            "text" : " ¬©",
            "bidi-level" : 0,
            "gravity" : "south",
            "language" : "en-us",
            "script" : "common",
            "font" : {
              "description" : "Cantarell 11",
              "checksum" : "5bcb6ee14ee9d210b2e91d643de1fe456e9d1aea770983fdb05951545efebbe2",
              "variations" : {
                "wght" : 0
              },
              "matrix" : [
                1,
                -0,
                -0,
                1,
                0,
                0
              ]
            },
            "flags" : 0,
            "y-offset" : 0,
            "start-x-offset" : 0,
            "end-x-offset" : 0,
            "glyphs" : [
              {
                "glyph" : 1109,
                "width" : 3072,
                "is-cluster-start" : true,
                "log-cluster" : 0
              },
              {
                "glyph" : 1182,
                "width" : 13312,
                "is-cluster-start" : true,
                "log-cluster" : 1
              }
            ]
          }
        ]
      }
    ]
  }
}
