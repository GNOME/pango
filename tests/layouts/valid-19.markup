# various Emoji segmentation cases
 a⃠ 😊︎ 😊️ 🇩🇪 ✊ ✋🏾 0 # 🏴󠁵󠁳󠁣󠁡󠁿 ©
