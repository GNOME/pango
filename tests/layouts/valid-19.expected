 a⃠ 😊︎ 😊️ 🇩🇪 ✊ ✋🏾 0 # 🏴󠁵󠁳󠁣󠁡󠁿 ©

--- parameters

wrapped: 0
ellipsized: 0
lines: 2

--- attributes

range 0 2147483647

--- directions

0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 

--- cursor positions

0(0) 1(0) 5(0) 6(0) 13(0) 14(0) 21(0) 22(0) 30(0) 31(0) 34(0) 35(0) 42(0) 43(0) 44(0) 45(0) 46(0) 47(0) 71(0) 72(0) 72(1) 75(0) 

--- lines

i=1, index=0, paragraph-start=1, dir=ltr ' a⃠ 😊︎ 😊️ 🇩🇪 ✊ ✋🏾 0 # 🏴󠁵󠁳󠁣󠁡󠁿 ©
'
i=2, index=75, paragraph-start=1, dir=ltr ''

--- runs

i=1, index=0, chars=4, level=0, gravity=south, flags=0, font=OMITTED, script=latin, language=en-us, ' a⃠ '
i=2, index=6, chars=3, level=0, gravity=south, flags=0, font=OMITTED, script=latin, language=en-us, '😊︎ '
i=3, index=14, chars=2, level=0, gravity=south, flags=0, font=OMITTED, script=latin, language=en-us, '😊️'
i=4, index=21, chars=1, level=0, gravity=south, flags=0, font=OMITTED, script=latin, language=en-us, ' '
i=5, index=22, chars=2, level=0, gravity=south, flags=0, font=OMITTED, script=latin, language=en-us, '🇩🇪'
i=6, index=30, chars=1, level=0, gravity=south, flags=0, font=OMITTED, script=latin, language=en-us, ' '
i=7, index=31, chars=1, level=0, gravity=south, flags=0, font=OMITTED, script=latin, language=en-us, '✊'
i=8, index=34, chars=1, level=0, gravity=south, flags=0, font=OMITTED, script=latin, language=en-us, ' '
i=9, index=35, chars=2, level=0, gravity=south, flags=0, font=OMITTED, script=latin, language=en-us, '✋🏾'
i=10, index=42, chars=5, level=0, gravity=south, flags=0, font=OMITTED, script=latin, language=en-us, ' 0 # '
i=11, index=47, chars=6, level=0, gravity=south, flags=0, font=OMITTED, script=latin, language=en-us, '🏴󠁵󠁳󠁣󠁡󠁿'
i=12, index=71, chars=2, level=0, gravity=south, flags=0, font=OMITTED, script=latin, language=en-us, ' ©'
i=13, index=74, no run, line end
i=14, index=75, no run, line end
